service/src/main/java/com/epam/esm/service/impl/TagServiceImpl.java

```diff
@@ -1,47 +1,55 @@
 package com.epam.esm.service.impl;
 
-import com.epam.esm.dao.entity.Tag;
-import com.epam.esm.dao.dao.impl.TagDaoImpl;
+import com.epam.esm.dao.dao.TagDao;
+import com.epam.esm.model.entity.Tag;
 import com.epam.esm.service.CommonService;
 import com.epam.esm.service.TagService;
+import com.epam.esm.service.exeption.RecourseExistException;
+import com.epam.esm.service.exeption.RecourseNotExistException;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
 
 import java.util.Optional;
 
 @Service
 public class TagServiceImpl implements CommonService<Tag>, TagService {
+    private final TagDao tagDao;
 
-    private final TagDaoImpl tagDao;
 
     @Autowired
-    public TagServiceImpl(TagDaoImpl tagDao) {
+    public TagServiceImpl(TagDao tagDao) {
         this.tagDao = tagDao;
     }
 
     @Override
-    public Tag add(Tag tag) {
-        return null;
+    @Transactional
+    public Tag add(Tag tag) throws RecourseExistException {
```

Comment:
Please **use DTO** objects for tags, certificates and other objects that you store in a database. The object that you store in a db and the object that your REST-service provides should be tightly tied

Reply:
@valevaty98 done